src/els_text_synchronization.erl

```diff
@@ -37,15 +37,33 @@ did_close(_Params) -> ok.
 
 -spec generate_diagnostics(uri()) -> ok.
 generate_diagnostics(Uri) ->
+  CDiagnostics = sync_diagnostics(Uri),
+  async_diagnostics(Uri, CDiagnostics).
```

Comment:
Why provide the compiler diagnostics here as well when they are already being sent as a notification in the synchronous diagnostics?

Reply:
We have to pass the compiler diagnostics down or else we loose them once the Dialyzer and Elvis diagnostics are reported. 

Take the following scenario as an example:
Given a module with a compiler error and an Elvis warning
When the diagnostics are run
Then the compiler error disappears once the Elvis warning comes in

By passing down the compiler error to the spawned process, the compiler errors are reported together with any Dialyzer and Elvis warnings and therefor the "compiler warnings are overwritten" when those notifications arrive. 

(I believe this also answers the question above in regards to generalizing this unless I misunderstood you?)

Reply:
The question above is more about having a more generic approach to the diagnostics. 

Something like this:
```
-spec generate_diagnostics(uri()) -> ok.
generate_diagnostics(Uri) ->
  Sync        = [els_compiler_diagnostics],
  Diagnostics = diagnostics(Uri, Sync, []),
  maybe_compile_and_load(Uri, Diagnostics),

  Async = [ els_dialyzer_diagnostics
          , els_elvis_diagnostics
          ],
  erlang:spawn(?MODULE, diagnostics, [Uri, Async, Diagnostics]),
  ok.

-spec diagnostics(uri(), [module()], [diagnostic()]) -> [diagnostic()].
diagnostics(Uri, Modules, Previous) ->
  Diagnostics = [apply(M, diagnostics, [Uri]) || M <- Modules],
  AllDiagnostics = lists:append([Previous | Diagnostics]),
  Method = <<"textDocument/publishDiagnostics">>,
  Params  = #{ uri         => Uri
             , diagnostics => AllDiagnostics
             },
  els_server:send_notification(Method, Params),
  AllDiagnostics.
```

Reply:
Ok cool, I see what you mean 👍 I'll fix it up and follow it with a squash to tidy things up.

I've noticed a significant improvement in my work flow with code_reload which is a lot more responsive now. And with some more fixes that we'll have to do to get Dialyzer in a happy state (it's missing a lot of type spec errors right now) Dialyzer will definitely become the slowest one in the diagnostics flow. 

So we might eventually even want to consider sending a message for each step and do the Elvis step before Dialyzer (I'm gonna make an assumption here and say for many this would be more important to get quick feedback on or perhaps it's just my personal view) but I'm not very familiar with the overhead and impact of sending multiple messages to the IDE to handle. Any thoughts on this?

Reply:
> So we might eventually even want to consider sending a message for each step

This could definitely be the way to go in the future, although I quite like the fast feedback vs. the slow feedback we have now.

>  ... but I'm not very familiar with the overhead and impact of sending multiple messages to the IDE to handle.

I guess this depends on how each IDE handles these messages. My experience with Emacs in this regard is that if there are too many (over a 100) errors or warnings to be shown, then interactions become slow.

Reply:
In the haskell language server, we cache the current diagnostics set for a file on the server, and update it based in `(Uri, diagnostics source)`. Whenever it changes we send to the client.

Also, we sort the diagnostics in order of occurrence in the file, and limit it to 100 max (a workspace config).

Further, if the compiler reports errors ( not warnings), we suppress diagnostics from other sources.

Reply:
@alanz Yeah we just discussed a similar approach but using a gen_server instead :) I'll start the work on this (probably on the weekend). This should create a much cleaner approach and especially for Dialyzer diagnostics, this should result in a smoother experience to the user :) (With some more changes that I'm working on, Dialyzer will likely take even longer...)

In that work I'll also include some tests for Dialyzer since all of this work isn't tested in a test suite other than "verifying that it works in a few different projects before creating a PR".